docs/proposals/koordlet/20221018-koordlet-support-cgroups-v2.md

```diff
@@ -0,0 +1,517 @@
+	ReadCPUAcctStat(parentDir string) (*CPUAcctStat, error)
+	ReadCPUStat(parentDir string) (*CPUStat, error)
+	ReadMemoryLimit(parentDir string) (int64, error)
+	ReadMemoryStat(parentDir string) (*MemoryStat, error)
+}
+```
+
+CgroupReader Implementations:
+
+```go
+func NewCgroupReader() CgroupReader {
+	if sysutil.Conf.UseCgroupV2 {
+		return &CgroupV2Reader{}
+	}
+	return &CgroupV1Reader{}
+}
+
+// for cgroups-v1
+type CgroupV1Reader struct{}
+
+func (r *CgroupV1Reader) ReadCPUQuota(parentDir string) (int64, error) {
+	s, err := sysutil.CgroupFileRead(parentDir, systemresource.CPUCFSQuota)
+	if err != nil {
+		return -1, fmt.Errorf("cannot read cgroup file, err: %v", err)
+	}
+	// content: "-1", "100000", ...
+	v, err := strconv.ParseInt(s, 10, 64)
+	if err != nil {
+		return -1, fmt.Errorf("cannot parse cgroup value %s, err: %v", s, err)
+	}
+	return v, nil
+}
+
+// for cgroups-v2
+type CgroupV2Reader struct{}
+
+func (r *CgroupV2Reader) ReadCPUQuota(parentDir string) (int64, error) {
+	s, err := sysutil.CgroupFileRead(parentDir, systemresource.CPUMaxV2)
+	if err != nil {
+		return -1, fmt.Errorf("cannot read cgroup file, err: %v", err)
+	}
+	// content: "max 100000", "100000 100000"
+	ss := strings.Fields(s)
+	if len(ss) != 2 {
+		return -1, fmt.Errorf("cannot parse cgroup value %s, err: invalid pattern", s)
+	}
+	if ss[0] == sysutil.CgroupMaxSymbolStr {
+		return -1, nil
+	}
+	v, err := strconv.ParseInt(ss[0], 10, 64)
+	if err != nil {
+		return -1, fmt.Errorf("cannot parse cgroup value %s, err: %v", s, err)
+	}
+	return v, nil
+}
+```
+
+#### Example: QoS Manager
+
+Here the CPU Burst Plugin scales container's cfs quota via the Resource Executor module:
+
+```go
+// CPUBurst Plugin
+func (p *Plugin) applyContainerCFSQuota(podMeta *statesinformer.PodMeta, containerStat *corev1.ContainerStatus,
+    curContainerCFS, deltaContainerCFS int64) error {
+    // get current pod-level cfs quota
+    curPodCFS, err := p.Executor().ReadCPUQuota(podMeta.CgroupDir)
+    if err != nil {
+        return err
+    }
+
+	if deltaContainerCFS > 0 { // cfs scale up, order: pod->container
+		if curPodCFS > 0 { // for pod-level; no need to adjust pod cpu.cfs_quota if it is already -1
+			targetPodCFS := curPodCFS + deltaContainerCFS
+			podCFSValStr := strconv.FormatInt(targetPodCFS, 10)
+			updater, err := executor.DefaultCgroupUpdaterFactory.New(systemresource.CPUCFSQuota, podDir, podCFSValStr)
+			if err != nil {
+			    return err
+			}
+			_, err := p.Executor().Update(true, updater)
+			if err != nil {
+			    return err
+			}
+		}
+
+		// for container-level
+	    containerDir, err := util.GetContainerCgroupPathWithKube(podMeta.CgroupDir, containerStat)
+	    if err != nil {
+		    return err
+	    }
+		targetContainerCFS := curContainerCFS + deltaContainerCFS
+		containerCFSValStr := strconv.FormatInt(targetContainerCFS, 10)
+		updater, err := executor.DefaultCgroupUpdaterFactory.New(systemresource.CPUCFSQuota, containerDir)
```

Comment:
take memory.min as an example, dose plugin need to do this?
```
if host is cgropu v1
    executor.DefaultCgroupUpdaterFactory.New(systemresource.MemMin, ...)
elseif host is cgropu v2
    executor.DefaultCgroupUpdaterFactory.New(systemresource.MemMinV2, ..)
```

Reply:
Fixed.  
Now we use `DefaultCgroupUpdaterFactory.New(t ResourceType, parentDir string, value string)`. It will return the resource updater adaptive to current cgroups version. e.g. returns the updater of `cpu.cfs_quota_us` for v1, and returns the updater of `cpu.max` for v2.